### ISSUE TITLE
Ambiguous definitions with DataArrays on v0.4

### ISSUE THREAD

**Issue by anriseth** [JuliaOpt/JuMP.jl#563]
With Julia 0.4..., DataArrays 0.2.18 and JuMP 0.10.0:
```julia
julia> using DataArrays, JuMP
WARNING: New definition 
    .+(AbstractArray{#S<:Union{JuMP.GenericQuadExpr{Float64, JuMP.Variable}, JuMP.GenericNormExpr{2, Float64, JuMP.Variable}, JuMP.AbstractJuMPScalar, JuMP.GenericNorm{P<:Any, Float64, JuMP.Variable}, JuMP.GenericAffExpr}, #N<:Any}, AbstractArray{#T<:Number, #N<:Any}) at /home/asbjorn/.julia/v0.4/JuMP/src/operators.jl:565
is ambiguous with: 
    .+(Union{DataArrays.PooledDataArray, DataArrays.DataArray}, AbstractArray...) at /home/asbjorn/.julia/v0.4/DataArrays/src/broadcast.jl:284.
To fix, define 
    .+(Union{DataArrays.DataArray{#S<:Union{JuMP.GenericQuadExpr{Float64, JuMP.Variable}, JuMP.GenericNormExpr{2, Float64, JuMP.Variable}, JuMP.AbstractJuMPScalar, JuMP.GenericNorm{P<:Any, Float64, JuMP.Variable}, JuMP.GenericAffExpr}, N<:Any}, DataArrays.PooledDataArray{#S<:Union{JuMP.GenericQuadExpr{Float64, JuMP.Variable}, JuMP.GenericNormExpr{2, Float64, JuMP.Variable}, JuMP.AbstractJuMPScalar, JuMP.GenericNorm{P<:Any, Float64, JuMP.Variable}, JuMP.GenericAffExpr}, R<:Integer, N<:Any}}, AbstractArray{#T<:Number, N<:Any})
before the new definition.

etc.
```

I don't know if this should be fixed in JuMP or DataArrays?

**Comment by joehuchette** [comment]
Hmm, that's no good. It's likely that both methods are overly broad. Thanks for reporting, I should have an easy fix.

**Comment by anriseth** [comment]
Great. Just to clarify, there are  more of these conflict warnings that
I've snipped out. I assume you can reproduce the error to see them all (I'm
not by a computer with Julia atm ...)

**Comment by IainNZ** [comment]
I get these on 0.3 on master:

```
IAINMAC:JuMP idunning$ jl test/runtests.jl
Warning: New definition
    .*(Union(SparseMatrixCSC{S<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),Ti<:Integer},Array{S<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),N}),Union(SparseMatrixCSC{T<:Number,Ti<:Integer},Array{T<:Number,N})) at /Users/idunning/.julia/v0.3/JuMP/src/operators.jl:568
is ambiguous with:
    .*(SparseMatrixCSC{Tv,Ti},SparseMatrixCSC{Tv,Ti}) at sparse/sparsematrix.jl:554.
To fix, define
    .*(SparseMatrixCSC{_<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),_<:Integer},SparseMatrixCSC{_<:Number,_<:Integer})
before the new definition.
Warning: New definition
    .*(Union(SparseMatrixCSC{S<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),Ti<:Integer},Array{S<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),N}),Union(Array{T<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),N},SparseMatrixCSC{T<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),Ti<:Integer})) at /Users/idunning/.julia/v0.3/JuMP/src/operators.jl:579
is ambiguous with:
    .*(SparseMatrixCSC{Tv,Ti},SparseMatrixCSC{Tv,Ti}) at sparse/sparsematrix.jl:554.
To fix, define
    .*(SparseMatrixCSC{_<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),_<:Integer},SparseMatrixCSC{_<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),_<:Integer})
before the new definition.
Warning: New definition
    .*(Union(Array{S<:Number,N},SparseMatrixCSC{S<:Number,Ti<:Integer}),Union(Array{T<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),N},SparseMatrixCSC{T<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),Ti<:Integer})) at /Users/idunning/.julia/v0.3/JuMP/src/operators.jl:587
is ambiguous with:
    .*(SparseMatrixCSC{Tv,Ti},SparseMatrixCSC{Tv,Ti}) at sparse/sparsematrix.jl:554.
To fix, define
    .*(SparseMatrixCSC{_<:Number,_<:Integer},SparseMatrixCSC{_<:Union(GenericQuadExpr{Float64,Variable},GenericAffExpr{CoefType,VarType},GenericNorm{P,Float64,Variable},GenericNormExpr{2,Float64,Variable},AbstractJuMPScalar),_<:Integer})
before the new definition.
```

**Comment by joehuchette** [comment]
:(

**Comment by joehuchette** [comment]
e3128857268be26546db4d8add808a1a5e84bb72

**Comment by IainNZ** [comment]
I'm still getting it, is that expected? Can you reproduce this?

**Comment by joehuchette** [comment]
I fixed it, then broke it, then fixed it again in d791cf249c083936b4b593c1cacfa9934f032013

**Comment by IainNZ** [comment]
Yep seems good now

**Comment by mlubin** [comment]
If it's resolved we should tag 0.10.1.

**Comment by joehuchette** [comment]
Sure, or we could leave it for a day or two and see if anything else shakes out.

**Comment by mlubin** [comment]
Okay with me

**Comment by mlubin** [comment]
Good thing we waited. https://github.com/JuliaLang/julia/commit/2984945c521fc70ed37b4e209d375dc2908fbdee just broke JuMP on 0.4

**Comment by mlubin** [comment]
Should be fixed on 0.10.1
